nixos/modules/installer/netboot/netboot.nix

```diff
@@ -0,0 +1,97 @@
+# This module creates netboot media containing the given NixOS
+# configuration.
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+{
+  options = {
+
+    netboot.storeContents = mkOption {
+      example = literalExample "[ pkgs.stdenv ]";
+      description = ''
+        This option lists additional derivations to be included in the
+        Nix store in the generated netboot image.
+      '';
+    };
+
+  };
+
+  config = {
+
+    boot.loader.grub.version = 2;
+
+    # Don't build the GRUB menu builder script, since we don't need it
+    # here and it causes a cyclic dependency.
+    boot.loader.grub.enable = false;
+
+    boot.initrd.postMountCommands = ''
+      mkdir -p /mnt-root/nix/store
+      mount -t squashfs /nix-store.squashfs /mnt-root/nix/store
+    '';
+
+    # !!! Hack - attributes expected by other modules.
+    system.boot.loader.kernelFile = "bzImage";
+    environment.systemPackages = [ pkgs.grub2 pkgs.grub2_efi pkgs.syslinux ];
+
+    boot.consoleLogLevel = mkDefault 7;
+
+    fileSystems."/" =
+      { fsType = "tmpfs";
+        options = [ "mode=0755" ];
+      };
+
+    boot.initrd.availableKernelModules = [ "squashfs" ];
+
+    boot.blacklistedKernelModules = [ "nouveau" ];
+
+    boot.initrd.kernelModules = [ "loop" ];
+
+    # Closures to be copied to the Nix store, namely the init
+    # script and the top-level system configuration directory.
+   netboot.storeContents =
+      [ config.system.build.toplevel ];
+
+    # Create the squashfs image that contains the Nix store.
+    system.build.squashfsStore = import ../../../lib/make-squashfs.nix {
+      inherit (pkgs) stdenv squashfsTools perl pathsFromGraph;
+      storeContents = config.netboot.storeContents;
+    };
+
+
+    # Create the initrd
+    system.build.netbootRamdisk = pkgs.makeInitrd {
+      inherit (config.boot.initrd) compressor prepend;
+
```

Comment:
@dezgeg, The example code you gave fails (assuming I copied it correctly...).

I made this change:
```diff
diff --git a/nixos/modules/installer/netboot/netboot.nix b/nixos/modules/installer/netboot/netboot.nix
index 41f79eb..7080b52 100644
--- a/nixos/modules/installer/netboot/netboot.nix
+++ b/nixos/modules/installer/netboot/netboot.nix
@@ -62,13 +62,11 @@ with lib;

     # Create the initrd
     system.build.netbootRamdisk = pkgs.makeInitrd {
-      inherit (config.boot.initrd) compressor prepend;
+      inherit (config.boot.initrd) compressor;
+      prepend = config.system.build.initialRamdisk;

       contents =
-        [ { object = config.system.build.bootStage1;
-            symlink = "/init";
-          }
-          { object = config.system.build.squashfsStore;
+        [ { object = config.system.build.squashfsStore;
             symlink = "/nix-store.squashfs";
           }
         ];
```

And got this error:
```
these derivations will be built:
  /nix/store/7h771qyvq24b1iz7233vb6wfrynfnmlp-initrd.drv
  /nix/store/fdmxgaw0rcjis4az63jr24q9vc8vfgx3-netboot.drv
building path(s) ‘/nix/store/pn2rr4wksz3f790hmnx0cpmsf0rwmii5-initrd’
cat: /nix/store/fdl35wv4i960fmx8yrj9ia971wpv1nm2-initrd: Is a directory
builder for ‘/nix/store/7h771qyvq24b1iz7233vb6wfrynfnmlp-initrd.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/fdmxgaw0rcjis4az63jr24q9vc8vfgx3-netboot.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/fdmxgaw0rcjis4az63jr24q9vc8vfgx3-netboot.drv’ failed
```

Something needs a trailing ```/initrd``` to find the file inside the directory.